domain/runaway.go

```diff
@@ -0,0 +1,525 @@
+	if err != nil {
+		return err
+	}
+	err = do.handleRemoveStaleRunawayWatch(records[0])
+	return err
+}
+
+func (do *Domain) runawayRecordFlushLoop() {
+	defer util.Recover(metrics.LabelDomain, "runawayRecordFlushLoop", nil, false)
+
+	// this times is used to batch flushing rocords, with 1s duration,
+	// we can guarantee a watch record can be seen by the user within 1s.
+	runawayRecordFluashTimer := time.NewTimer(runawayRecordFlushInterval)
+	runawayRecordGCTicker := time.NewTicker(runawayRecordGCInterval)
+	failpoint.Inject("FastRunawayGC", func() {
+		runawayRecordFluashTimer.Stop()
+		runawayRecordGCTicker.Stop()
+		runawayRecordFluashTimer = time.NewTimer(time.Millisecond * 50)
+		runawayRecordGCTicker = time.NewTicker(time.Millisecond * 200)
+	})
+
+	fired := false
+	recordCh := do.RunawayManager().RunawayRecordChan()
+	quarantineRecordCh := do.RunawayManager().QuarantineRecordChan()
+	staleQuarantineRecordCh := do.RunawayManager().StaleQuarantineRecordChan()
+	flushThrehold := do.runawayManager.FlushThreshold()
+	records := make([]*resourcegroup.RunawayRecord, 0, flushThrehold)
+
+	flushRunawayRecords := func() {
+		if len(records) == 0 {
+			return
+		}
+		sql, params := resourcegroup.GenRunawayQueriesStmt(records)
+		if _, err := do.execRestrictedSQL(sql, params); err != nil {
+			logutil.BgLogger().Error("flush runaway records failed", zap.Error(err), zap.Int("count", len(records)))
+		}
+		records = records[:0]
+	}
+
+	for {
+		select {
+		case <-do.exit:
+			return
+		case <-runawayRecordFluashTimer.C:
+			flushRunawayRecords()
+			fired = true
+		case r := <-recordCh:
+			records = append(records, r)
+			failpoint.Inject("FastRunawayGC", func() {
+				flushRunawayRecords()
+			})
+			if len(records) >= flushThrehold {
+				flushRunawayRecords()
+			} else if fired {
+				fired = false
+				// meet a new record, reset the timer.
+				runawayRecordFluashTimer.Reset(runawayRecordFlushInterval)
+			}
+		case <-runawayRecordGCTicker.C:
+			go do.deleteExpiredRows("tidb_runaway_queries", "time", runawayRecordExpiredDuration)
+		case r := <-quarantineRecordCh:
+			go func() {
+				err := do.handleRunawayWatch(r)
+				if err != nil {
+					logutil.BgLogger().Error("add runaway watch", zap.Error(err))
+				}
+			}()
+		case r := <-staleQuarantineRecordCh:
+			go func() {
+				for i := 0; i < 3; i++ {
+					err := do.handleRemoveStaleRunawayWatch(r)
+					if err == nil {
+						break
+					}
+					logutil.BgLogger().Error("remove stale runaway watch", zap.Error(err))
+					time.Sleep(1000 * time.Millisecond)
+				}
+			}()
+		}
+	}
+}
+
+func (do *Domain) handleRunawayWatch(record *resourcegroup.QuarantineRecord) error {
+	se, err := do.sysSessionPool.Get()
+	defer func() {
+		do.sysSessionPool.Put(se)
+	}()
+	if err != nil {
+		return errors.Annotate(err, "get session failed")
+	}
+	exec, _ := se.(sqlexec.SQLExecutor)
+	ctx := kv.WithInternalSourceType(context.Background(), kv.InternalTxnStats)
+	_, err = exec.ExecuteInternal(ctx, "BEGIN")
```

Comment:
why ignore checking this error

Reply:
In the design, we want to return the ID of the added watch record to the user.